File: templates/problems/10_backtracking/01_subsets/typescript/solution.ts
/**
 * Generate all subsets using backtracking.
 * Time: O(n * 2^n), Space: O(n) recursion depth
 */
export function solveSubsets(nums: number[]): number[][] {
    const result: number[][] = [];

    const backtrack = (start: number, current: number[]): void => {
        result.push([...current]);
        for (let i = start; i < nums.length; i++) {
            current.push(nums[i]);
            backtrack(i + 1, current);
            current.pop();
        }
    };

    backtrack(0, []);
    return result;
}

File: templates/problems/10_backtracking/01_subsets/python/solution.py
def solve_subsets(nums: list[int]) -> list[list[int]]:
    """Generate all subsets using backtracking.

    Time: O(n * 2^n), Space: O(n) recursion depth
    """
    result: list[list[int]] = []

    def backtrack(start: int, current: list[int]) -> None:
        result.append(list(current))
        for i in range(start, len(nums)):
            current.append(nums[i])
            backtrack(i + 1, current)
            current.pop()

    backtrack(0, [])
    return result
